### ISSUE TITLE
Proposal: no-redeclare reports /*globals foo*/ comments if it's redeclaration

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#11370]
**What rule do you want to change?**

- [no-redeclare] rule

**Does this change cause the rule to produce more or fewer warnings?**

- more warnings

**How will the change be implemented? (New option, new default behavior, etc.)?**

- I want to change the default behavior. But I'm OK if it's behind an option.

**Please provide some example code that this change will affect:**

config:

```yml
globals:
  window: readable
rules:
  no-redeclare:
    - error
    - builtinGlobals: true
```

code:

```js
/*globals window */
```

**What does the rule currently do for this code?**

- Does nothing.

**What will the rule do after it's changed?**

- Reports the `window` in `/*globals*/` comment because it's declared already.

For a reference, [no-unused-vars] rule reports unused variables that `/*globals*/` comment defined. For example:

> ```js
> /*globals foo */ //ERROR: 'foo' is defined, but never used.
> ```
>
> https://eslint.org/demo/#eyJ0ZXh0IjoiLypnbG9iYWxzIGZvbyAqLyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6OCwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJuby11bnVzZWQtdmFycyI6Mn0sImVudiI6e319fQ==

Therefore, this change makes consistent between [no-unused-vars] rule and [no-redeclare] rule.

**Are you willing to submit a pull request to implement this change?**

Yes, I will do.

[no-redeclare]: https://eslint.org/docs/rules/no-redeclare
[no-unused-vars]: https://eslint.org/docs/rules/no-unused-vars

**Comment by mysticatea** [comment]
I added `core` label because of a few changes required to implement this.

**Comment by not-an-aardvark** [comment]
Could you elaborate on what changes would be needed to core to implement this?

**Comment by mysticatea** [comment]
Sure. The change will happen in https://github.com/eslint/eslint/blob/0a3c3ff1d91e8f39943efc4a7d2bf6927d68d37e/lib/linter.js#L72

It needs to keep all locations rather than overriding. `variable.eslintExplicitGlobalComment` will be an array of `globals` comments, and `variable.eslintConfigSetting` or something like is needed to know eslintrc's setting.

**Comment by not-an-aardvark** [comment]
Thanks for elaborating. I think this sounds like a good idea overall, but it would be useful to have a more concrete proposal for what the breaking changes are. (For example, would there be two variables with the same name in the global `scope`? Or would additional metadata be added to a single variable?

Before https://github.com/eslint/eslint/commit/0a3c3ff1d91e8f39943efc4a7d2bf6927d68d37e, I think `/* global */` comments and configured globals with the same name would create separate variables, but this was changed to ensure `/* global Foo: off */` would work correctly.

After this change, would the following create a `no-redeclare` error since it changes the config for `Foo`?

```json
{
    "globals": {
        "Foo": "writable"
    }
}
```

```js
// Change `Foo` from "writable" to "readable"
/* global Foo */
```

**Comment by mysticatea** [comment]
Thank you for the feedback!

> For example, would there be two variables with the same name in the global `scope`? Or would additional metadata be added to a single variable?

This proposal will do:

- Modify metadata in existing variable objects.  
  (makes `variable.eslintExplicitGlobalComment` to an array)
- Add a metadata into existing variable objects.
  (`variable.eslintConfigFileSetting` which is the configuration in `eslintrc` file)

> `/* global Foo: off */`

My preference is that `no-redeclare` rule will ignore such a removed declarations because it's a corner case and ECMAScript doesn't have the analogy of that.

> Change `Foo` from "writable" to "readable"

My preference is that `no-redeclare` rule will warn it because it's similar to the redeclaration with `let` and `const`.

```ts
const Foo = 0
let Foo = 1 // 'Foo' is defined already.
```

**Comment by not-an-aardvark** [comment]
Changing `variable.eslintExplicitGlobalComment` to an array would probably break all users of the `eslintExplicitGlobalComment` property. I wasn't sure how many people are actually using `eslintExplicitGlobalComment`, so I did a [GitHub search](https://github.com/search?q=eslintExplicitGlobalComment+-filename%3Ano-unused-vars.js+-filename%3Aeslint.js+-filename%3Alinter.js+-filename%3A*.svn-base&type=Code). I couldn't find any plugins using it, so it's probably fine to change it to an array as a breaking change.

However, would we want to name the property `variable.eslintExplicitGlobalComments` and deprecate or remove `variable.eslintExplicitGlobalComment`?

**Comment by mysticatea** [comment]
> However, would we want to name the property `variable.eslintExplicitGlobalComments` and deprecate or remove `variable.eslintExplicitGlobalComment`?

Sounds good to me.

**Comment by nzakas** [comment]
We should do an RFC for this.

**Comment by not-an-aardvark** [comment]
This change was accepted in [today's TSC meeting](https://github.com/eslint/tsc-meetings/blob/3e40976a2ebb9481405d4c3a358bc3191ecd234a/notes/2019/2019-02-28.md#fix-no-var-fixed-to-incorrect-code-fixes-11441), but there was agreement that an RFC would be useful to decide on the precise changes to core.
